### ISSUE TITLE
Stream URL getting "Connection Reset by Peer"

### ISSUE THREAD

**Issue by abettadapur** [simon-weber/gmusicapi#286]
Hey. I am using VLC to play URLs returned from getStreamUrl(). I only get the URL when I am about to play it, so as to avoid the URL expiring. However, periodically, VLC will skip the track and complain that the "Connection was Reset by Peer." Do you have any insight on this issue? Is Google closing the connection?

**Comment by simon-weber** [comment]
This issue hasn't been updated in a while, so I'm closing it. Feel free to reopen if you're still having trouble.

**Comment by simon-weber** [comment]
@vmurillo seems to be able to recreate this (copied from #362):

> Whenever I trye to play a podcast, after a small period of time I got this console error:
> [00007f3940c09938] core access error: read error: Connection reset by peer
> I am using VLC 2.2.1buildd@lgw01-37.buildd Jul 20 2015 23:42:41 under Ubuntu 14.04

I wonder if it's something specific to how vlc reads the stream? If it's doing something like requesting it multiple times and using range headers, it might be running into what @thebigmunch described earlier.

**Comment by kettenbach-it** [comment]
Hi,

I think this problem with Logitech Squeezeboxserver comes from the TCP-RST as well:
https://github.com/Logitech/slimserver/issues/130#issuecomment-279213259

For the Squeezeboxserver this is really a problem: the Google Music Plugin is almost useless afterwards

**Comment by simon-weber** [comment]
Interesting. Unfortunately, I don't have a good way of getting that to the Music team aside from lots of people independently reporting it.

**Comment by kettenbach-it** [comment]
I don't think it would make sense to report it to google, since this is not an official API.

How about that: the URL that Google returns on playing music contains the expiry date of the URL. Wouldn't it make sense to call that URL shortly before it's expiry to refresh it?

(I don't know anything about the way this API works, so please excuse any stupid ideas).

Maybe we have to check how gmusicproxy handles this.

**Comment by thebigmunch** [comment]
Your solution would be something that happens outside of gmusicapi. As far as gmusicapi's involvement, it makes the call to get the URL and gives it to the caller. It's up to the caller to do whatever it wants with it. So, you'd have to propose something like that to the application.

That being said, it also probably wouldn't solve anything, as the issue described here is that the connection is getting reset after streaming has already begun. This could still happen before the streaming ends even if the application gives itself the ~90 second buffer of reloading right before the expiry. And if you're talking about getting a new URL entirely (the only way to get a later expiry is to get a new URL), then the application would have to handle that.

There's nothing to be done about this in gmusicapi itself. And, as Simon said, there's little recourse in getting the Google Music team's input on it. But things have been done when many independent reports have been sent to the Google Music team in the past.

**Comment by kettenbach-it** [comment]
Your right.
I digged in to gmusicproxy as a refernce.
And I see, that this has to be handled by https://github.com/squeezebox-googlemusic

**Comment by MichaelCDormann** [comment]
Has anyone figured out a fix for this? Just found this Python module last night and was very excited about it. Sucks that when I had some stuff working I ran into this same issue.

**Comment by kettenbach-it** [comment]
I switched to Spotify

**Comment by simon-weber** [comment]
Maybe Google expects clients to buffer the stream locally for playback? That'd make if they're looking to minimize their number of open connections and should be easy enough to try out. I think their clients do this (I haven't confirmed it, but I know the web client will continue playing even if you lose connection in the middle of a song).

**Comment by thebigmunch** [comment]
The mobile client has an option to cache while streaming.

**Comment by MichaelCDormann** [comment]
Yea, I was thinking about possibly buffering the stream. That's probably what I'll have to do.

**Comment by MichaelCDormann** [comment]
So I used urllib to open the stream url and write it to a local mp3 file. After that it's simple enough to play the mp3 file, and you don't have to worry about the stream url expiring. Kind of an ugly way of "buffering" and I might look into doing it another way, but it works.

Example (segments from my working code):
```python
import urllib.request

stream_url = api.get_stream_url(song_id=self.song_id, quality=u'hi')

mp3file = urllib.request.urlopen(stream_url)
with open('./test.mp3', 'wb') as output:
    output.write(mp3file.read())

vlc.MediaPlayer('./test.mp3')
```

**Comment by sundown94** [comment]
Hi Everyone!
I think I got this working.  My configuration (after lots of trial and error) is Raspian Jessie, LMS 7.9.1 (logitechmediaserver_7.9.1~1489384180_arm.deb), gmusicapi v10.1.1, and the latest Google Music plugin (along with the HTTPS change to line 5 in it's protocolhandler.pm file).

Sorry if this is already old news, but I couldn't find this resolution documented anywhere.  I also posted a similar response over on Logitech/slimserver#130.
